### ISSUE TITLE
[Bug] Type-checking violations for lambda-style functional validators defined in IPython REPL omit source code

### ISSUE THREAD

**Issue by braniii** [beartype/beartype#123]
I really like the idea of functional validators. They make the source code much more robust while keeping it lean and clean. My only concern is that when the bear roars, the user has no idea what she/he did wrong. Probably, I am using beartype the wrong way.

Using a simple type the user gets an kind of helpful warning, e.g.,
```python
>>> import numpy as np
>>> from beartype import beartype
>>> from beartype.vale import Is
>>> from numpy.typing import NDArray
>>> from typing_extensions import Annotated
>>> @beartype
... def func(a: int) -> int:
...     return a
... 
>>> 
>>> func(5)
5
>>> func(5.)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<@beartype(__main__.func) at 0x7f6f2cb03940>", line 21, in func
  File "/home/user/anaconda3/envs/py3.8/lib/python3.8/site-packages/beartype/_decor/_error/errormain.py", line 301, in raise_pep_call_exception
    raise exception_cls(  # type: ignore[misc]
beartype.roar.BeartypeCallHintParamViolation: @beartyped func() parameter a=5.0 violates type hint <class 'int'>, as 5.0 not instance of int.
```
so it is clear, that an integer was expected. In this case it is trivial, but probably it would be helpful to also print the type of the input variable. So that it becomes clearer.
Using a validator these error are getting very cryptic. Lets define three super handy validators.
```python
>>> Float2DArray = Annotated[
...     NDArray[np.floating],
...     Is[lambda arr: arr.ndim == 2],
... ]
>>> FloatMatrix = Annotated[
...     Float2DArray,
...     Is[lambda arr: arr.shape[0] == arr.shape[1]],
... ]
>>> SymmetricFloatMatrix = Annotated[
...     FloatMatrix,
...     Is[lambda arr: np.allclose(arr, arr.T)],
... ]
>>> 
>>> @beartype
... def func(arr: SymmetricFloatMatrix) -> SymmetricFloatMatrix:
...     return arr
...
>>> 
>>> # this runs fine
>>> mat = np.zeros((4, 4))
>>> func(mat)
array([[0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.]])
```
So far, I am happy. But now parsing an array of wrong shape,
```python
>>> # not 2D
>>> mat = np.zeros((4, ))
>>> func(mat)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<@beartype(__main__.func) at 0x7f6f2f37e820>", line 53, in func
  File "/home/user/anaconda3/envs/py3.8/lib/python3.8/site-packages/beartype/_decor/_error/errormain.py", line 301, in raise_pep_call_exception
    raise exception_cls(  # type: ignore[misc]
beartype.roar.BeartypeCallHintParamViolation: @beartyped func() parameter arr="array([0., 0., 0., 0.])" violates type hint typing_extensions.Annotated[numpy.ndarray, Is[<lambda>], Is[<lambda>], Is[<lambda>]], as "array([0., 0., 0., 0.])" violates validator Is[<lambda>]:
    False == Is[<lambda>].
```
a non quadratic array (matrix), or
```python
>>> # not quadratic
>>> mat = np.zeros((4, 5))
>>> func(mat)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<@beartype(__main__.func) at 0x7f6f2f37e820>", line 53, in func
  File "/home/user/anaconda3/envs/py3.8/lib/python3.8/site-packages/beartype/_decor/_error/errormain.py", line 301, in raise_pep_call_exception
    raise exception_cls(  # type: ignore[misc]
beartype.roar.BeartypeCallHintParamViolation: @beartyped func() parameter arr="array([[0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0.],
    ...[ violates type hint typing_extensions.Annotated[numpy.ndarray, Is[<lambda>], Is[<lambda>], Is[<lambda>]], as "array([[0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0.],
    ...[ violates validator Is[<lambda>]:
    False == Is[<lambda>].
```
a non symmetric matrix,
```python
>>> # not symmetric
>>> mat = np.zeros((4, 4))
>>> mat[0, 1] = 2
>>> func(mat)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<@beartype(__main__.func) at 0x7f6f2f37e820>", line 53, in func
  File "/home/user/anaconda3/envs/py3.8/lib/python3.8/site-packages/beartype/_decor/_error/errormain.py", line 301, in raise_pep_call_exception
    raise exception_cls(  # type: ignore[misc]
beartype.roar.BeartypeCallHintParamViolation: @beartyped func() parameter arr="array([[0., 2., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0... violates type hint typing_extensions.Annotated[numpy.ndarray, Is[<lambda>], Is[<lambda>], Is[<lambda>]], as "array([[0., 2., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0... violates validator Is[<lambda>]:
    False == Is[<lambda>].
```
In each case I am getting errors that some `Is[<lambda>]` are not fulfiled. If I remember correctly, older versions of beartype did print the lambda functions code. Using a wrong type, the error is more helpful, e.g.,
```python
>>> func(5)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<@beartype(__main__.func) at 0x7f6f2f37e820>", line 53, in func
  File "/home/user/anaconda3/envs/py3.8/lib/python3.8/site-packages/beartype/_decor/_error/errormain.py", line 301, in raise_pep_call_exception
    raise exception_cls(  # type: ignore[misc]
beartype.roar.BeartypeCallHintParamViolation: @beartyped func() parameter arr=5 violates type hint typing_extensions.Annotated[numpy.ndarray, Is[<lambda>], Is[<lambda>], Is[<lambda>]], as 5 not instance of <protocol "numpy.ndarray">.
```
I would love to see a list with requirements that the variable needs to fulfill. So something like:

```python
beartype.roar.BeartypeCallHintParamViolation: @beartyped func() parameter arr="array([[0., 2., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0..
violates type hint SymmetricFloatMatrix:
       - numpy.typing.NDArray[numpy.floating]
       - lambda arr: arr.ndim == 2
       - lambda arr: arr.shape[0] == arr.shape[1]
       - lambda arr: np.allclose(arr, arr.T)
```

So back to the original question. Is is possible to improve readability of the errors?

**Comment by leycec** [comment]
> ...but probably it would be helpful to also print the type of the input variable.

**Yes ⨯ ∞.** Your wisdom exceeds mine yet again, @branii. Because I have the memory of a guppy, I've opened a [new feature request tracking this rapaciously perspicacious observation](https://github.com/beartype/beartype/issues/124).

Let's type this!

> In each case I am getting errors that some `Is[<lambda>]` are not fulfilled.

**Ah-ha!** I *love* fielding questions I know more than zero about. This could actually be one such question.

Let's talk lambdas. Recovering the human-readable source code underlying a lambda function object is **highly non-trivial** – like, mind-blowingly universe-destroyingly non-trivial. I devoted [a full 500-line submodule embedded deep within the perfidious bowels of the `beartype` codebase to the subject](https://github.com/beartype/beartype/blob/da1bff6376b317195d55dceb86fed2300cb2aad5/beartype/_util/func/utilfunccode.py#L248). While doing so, we actually discovered two significant issues in the **CPython** codebase itself. Because laziness prevails, I naturally forgot to submit upstream CPython issues. Bugs prevail... *yet again.*

For the dangerously incautious who wish to repeat my life mistakes, recovering the source code for a lambda requires decompiling that lambda's code object first into abstract syntax tree (AST) nodes and then into source code with a recursive tree walk across all AST nodes of the file defining that lambda using the [Gang of Four's (GoF's) infamous Visitor design](https://en.wikipedia.org/wiki/Visitor_pattern) filtering for the leaf AST node corresponding to that lambda.

Honestly? It's insane. Pertinently, the relevant keyword above for this issue is "file." This currently only applies to lambdas defined in actual files, because... *laziness prevails.* Recovering lambdas defined only in-memory (e.g., in an IPython REPL) is technically feasible, but even less trivial, because it requires use of the [mostly undocumented hidden `linecache` module](https://stackoverflow.com/a/41001920/2809027).

## What Are You Trying to Say Here?

What I'm trying but failing to say here is that all of your examples do in fact behave as expected for functional validators defined in **files**:

```
# In "Big_Big_braniii.py":
import numpy as np
from beartype import beartype
from beartype.vale import Is
from numpy.typing import NDArray
from typing import Annotated

Float2DArray = Annotated[
    NDArray[np.floating],
    Is[lambda arr: arr.ndim == 2],
]
FloatMatrix = Annotated[
    Float2DArray,
    Is[lambda arr: arr.shape[0] == arr.shape[1]],
]
SymmetricFloatMatrix = Annotated[
    FloatMatrix,
    Is[lambda arr: np.allclose(arr, arr.T)],
]

@beartype
def func(arr: SymmetricFloatMatrix) -> SymmetricFloatMatrix:
    return arr

mat = np.zeros((4,))
func(mat)
```

```bash
$ python3.10 Big_Big_braniii.py
Traceback (most recent call last):
  File "/home/leycec/tmp/Big_Big_braniii.py", line 27, in <module>
    func(mat)
  File "<@beartype(__main__.func) at 0x7ff4a9d23490>", line 53, in func
  File "/usr/lib/python3.10/site-packages/beartype/_decor/_error/errormain.py", line 301, in raise_pep_call_exception
    raise exception_cls(  # type: ignore[misc]
beartype.roar.BeartypeCallHintParamViolation: @beartyped func() parameter arr="array([0., 0., 0., 0.])" violates type hint typing.Annotated[numpy.ndarray[typing.Any, numpy.dtype[numpy.floating]], Is[lambda arr: arr.ndim == 2], Is[lambda arr: arr.shape[0] == arr.shape[1]], Is[lambda arr: np.allclose(arr, arr.T)]], as "array([0., 0., 0., 0.])" violates validator Is[lambda arr: arr.ndim == 2]:
    False == Is[lambda arr: arr.ndim == 2].
```

Of course, REPLs are not files; they're purely in-memory.

Because Python gonna Python, what you and everyone wants is sadly infeasible for the standard Python REPL (i.e., what you get when you run `python3` without arguments). Why? Because the standard Python REPL fails to register callables (*...yup, this includes lambda functions*) with `linecache`. So, the original source code can *never* be recovered. It's gone.

On the other hand, what you and everyone wants is absolutely feasible for smarter REPLs. This includes both the IPython REPL and [possibly Jupyter Notebooks](https://stackoverflow.com/a/65806216/2809027). Unlike the standard Python REPL, IPython at least registers callables with `linecache`.

## What You Gonna Do About Your Brokenness

We're gonna fix it! Basically, we need to make our low-level machinery that recovers source code from arbitrary callables `linecache`-aware. This part is mostly trivial, because `linecache` presents a [mostly sane dictionary API for this](https://stackoverflow.com/a/41001920/2809027).

The caveat is that **this will probably only work for IPython.** The standard Python REPL is functionally useless, while [Jupyter doesn't appear to integrate with `linecache`](https://stackoverflow.com/a/65806216/2809027).

What will @leycec do next? *Even he does not know.*

**Comment by braniii** [comment]
Thx a lot for this great answer. Sadly, this sounds way more complicated as expected. And to be honest, after reading your answer I feel like I've asked a stupid question :sweat_smile: But being lazy, there needs to be a simpler solution!? Would it be possible to add a second (optional) argument to `beartype.vale.Is` where one could specify an error string? This would be less elegant, but maybe easier to implement.

**Comment by braniii** [comment]
I guess was still sleeping while reading your answer, I missed the solution to my issue: concatenating the validators is quite close to what I was looking for. Thx :)

**Comment by braniii** [comment]
I guess was still sleeping while reading your answer, I missed the solution to my issue: concatenating the validators is quite close to what I was looking for. Thx :)

**Comment by leycec** [comment]
**You're most welcome.** Oh, and I couldn't quite decide which REPL you were using. Was that IPython or something less awesome? If IPython, the upcoming @beartype release (...so, `beartype 0.10.5` in the countably infinite `0.10.x` release cycle) will solve at least the ambiguous `Is[<lambda>]` issue for you by embedding the source code of functional validators defined in the IPython REPL within exception messages.

Until then, defining functional validators in actual files is the way to go. Zombie emoji... *just because.* :zombie: :zombie_man: :zombie_woman:

**Comment by braniii** [comment]
Oh, no :see_no_evil:  I was using the default Python terminal (probably 3.9.12)
>  If I remember correctly, older versions of beartype did print the lambda functions code. 

This sentence from my first post iis obviously bullshit. Normally, I never use a Python terminal together with beartype. But recently I did to convince my colleagues to start using beartype. And I thought it changed due to an update.

**Comment by leycec** [comment]
> I was using the default Python terminal

:scream: 

> This sentence from my first post iis obviously bullshit.

If I had a bitcoin for every time I myself admitted that, our cats would be sipping milk margaritas in Cancún.

<sup>You're awesome is what I'm saying.</sup>

> But recently I did to convince my colleagues to start using beartype.

First, I went :slightly_smiling_face: .

Then, I went :smiling_face_with_tear: after realizing that failed painfully and everyone tried awkwardly changing the subject to hypothetical storyboards for *Rick and Morty* Season 6. But not even cooler talk could dampen the blow.

It's almost like I was there, @braniii.

**Comment by dycw** [comment]
My $0.02 here: out of habit, I have my all validators defined as functions, e.g.

```python
def is_array_2d(arr):
    return arr.ndim == 2
```

so failures immediately pointed to by via function name.

Note that `beartype.vale` is pretty powerful, and some of the things you'd want to check can be implemented using native primitives. For example,

```python
NDim2 = IsAttr['ndim', IsEqual[2]]
```

works, or if you want to get fancy:


```python
class _NDim2(type):
    def __getitem__(cls, n):
        return IsAttr['ndim', IsEqual[n]]

NDim[1], NDim[2], ...
```

would all work too. HTH!

**Comment by leycec** [comment]
```python
class _NDimMeta(type):
    def __getitem__(cls, n):
        return IsAttr['ndim', IsEqual[n]]
```

Let us appreciate how undeniably hot that is. :sauna_man:

**Comment by braniii** [comment]
Ok this is next level. I was not aware that metaclasses can be so handy.

**Comment by TeamSpen210** [comment]
Not sure why you're using a metaclass there actually. The `NDim` object only exists to be subscripted, so you could just do:
```py
@object.__new__
class NDim:
    def __getitem__(self, n):
        return IsAttr['ndim', IsEqual[n]]
# or instead of the decorator:
NDim = NDim()
```

**Comment by leycec** [comment]
> Not sure why you're using a metaclass there actually.

Because... *it is hot.*

Seriously, though. Good catch – as always! KISS or GTFO. But what is this is shadow madness:

```python
@object.__new__
```

*furious Googling intensifies*

Relatedly, if we're maximizing terseness for all the lols:

```python
class NDim:
    def __class_getitem__(self, n):  # <-- PEP 560, yo
        return IsAttr['ndim', IsEqual[n]]
```

:muscle:

**Comment by leycec** [comment]
**So, it begins.** Kinda unsure if anyone actually cares about beartype validators in IPython, but let's just do this anyway. Silence, unborn future complainers! You will no longer have anything to complain about.

Incidentally, I'm pretty sure this emoji features a man feeding a rabid ferret: :man_feeding_baby: <sup>← *wut are you doing, man*</sup>

**Comment by leycec** [comment]
**Resolved by d0df91d.** This was surprisingly easy. Thanks for the save, [standard `linecache` module](https://docs.python.org/3/library/linecache.html). IPython REPL output or it didn't happen, so:

```python
$ ipython3
Python 3.8.13 (default, Mar 23 2022, 03:39:13) 
Type 'copyright', 'credits' or 'license' for more information
IPython 7.30.1 -- An enhanced Interactive Python. Type '?' for help.

[ins] In [1]: from beartype.vale import Is

[ins] In [2]: IsTrue = Is[lambda obj: True]

[ins] In [3]: repr(IsTrue)
Out[5]: 'Is[IsTrue = Is[lambda obj: True]\n]'
```

That's arguably not the best output. Okay, it's awful. But it's totally *not* our fault. <sup>I swear this on five fat Maine Coon cats.</sup> That's what IPython is erroneously caching with `linecache`. We blame Jupyter.

Still, anything beats `Is[<lambda>]`. By strenuous squinting (that threatens cataracts at age 45), at least you can *sorta* decipher what's going on. It's good enough for @beartype. This means it's good enough for you! :monocle_face:
